### ISSUE TITLE
OpenBlas Multi-Threading on Windows

### ISSUE THREAD

**Issue by HaoTang1** [dmlc/mxnet#937]
New to MxNet. I've set up and tested it on a 12-core windows machine, tried both the pre-build CPU-only package and building the package myself using CMake. 

I tried the sample project under example/notebooks/cifar10-recipe.ipynb. I changed the python code to CPU-only mode, and modified the preprocess_threads and MXNET_CPU_WORKER_NTHREADS to 12 since I have 12 cores, but the training speed ( [12] in cifar10-recipe.ipynb) is still really slow, and according to the hardware monitor it only loads one core to 100%, other cores are still in idle. 
```
INFO:root:Iter[0] Batch [50] Speed: 1.83 samples/sec
INFO:root:Iter[0] Batch [100] Speed: 1.77 samples/sec
INFO:root:Iter[0] Batch [150] Speed: 1.77 samples/sec
INFO:root:Iter[0] Batch [200] Speed: 1.78 samples/sec
INFO:root:Iter[0] Batch [250] Speed: 1.78 samples/sec
INFO:root:Iter[0] Batch [300] Speed: 1.78 samples/sec
INFO:root:Iter[0] Batch [350] Speed: 1.78 samples/sec
```

Are there any other settings related to multi-core utilization? Or is GPU based system the only way to go for large data set?

Thanks

**Comment by tqchen** [comment]
If compiled with OpenBLAS, the blas kernel should be able to enable parallelization and it is not related to ```MXNET_CPU_WORKER_NTHREADS ```. Though I do not know the case for window's OpenBLAS

**Comment by tqchen** [comment]
loop in @hjk41

**Comment by hjk41** [comment]
Try to set OPENBLAS_NUM_THREADS
This should enable OpenBLAS to use multiple threads to do computation

**Comment by HaoTang1** [comment]
@hjk41 
I tried to set OPENBLAS_NUM_THREADS to 12, still didn't change anything. 

In the OpenBLAS GitHub page I saw this in the documentation.
"If you compile this lib with USE_OPENMP=1, you should set OMP_NUM_THREADS environment variable. OpenBLAS ignores OPENBLAS_NUM_THREADS and GOTO_NUM_THREADS with USE_OPENMP=1."

Yes, I'm using OpenMP right now, so I changed OMP_NUM_THREADS to 12. But still, only one core is running at 100%.

**Comment by piiswrong** [comment]
I think this is a openblas issue. I wasn't able to get openblas running on multiple cores with octave either

**Comment by hjk41** [comment]
Let me check with OpenBLAS

On Wed, Dec 16, 2015 at 11:27 AM, Eric Junyuan Xie <notifications@github.com
> wrote:

> I think this is a openblas issue. I wasn't able to get openblas running on
> multiple cores with octave either
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dmlc/mxnet/issues/937#issuecomment-164978199>.
>



-- 
HONG Chuntao
System Research Group
Microsoft Research Asia

**Comment by hjk41** [comment]
Checked with OpenBLAS author, Windows binary distribution *DOES* support multi-threading. I also tried on my own machine, setting OPENBLAS_NUM_THREADS to 16 boosts the CPU utilization to 100% on my 16-core machine.

I think it depends on how the cifar example uses the library. If it does not use matrix multiplication much, you will not see much speedup when using multiple threads. I think for Cifar, most of the time should be spent on the convolution layers? Then it comes to how convolution is implemented on CPU. Can @antinucleon shed light on this?
